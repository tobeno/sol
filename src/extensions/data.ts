import { ClassWithText } from '../interfaces/base';
import { Constructor } from '../interfaces/util';
import { awaitPromiseSync } from '../utils/async';

export function WithData<T extends Constructor<ClassWithText>>(base: T) {
    return class Wrapped extends base {
        get data(): any {
            if (super.data) {
                return super.data;
            }

            return JSON.parse(this.text);
        }

        set data(value: any) {
            if (super.data) {
                super.data = value;

                return;
            }

            this.text = JSON.stringify(value, null, 2);
        }

        updateData(updater: (value: any) => any | Promise<any>) {
            const updated = awaitPromiseSync(updater(this.data));
            if (typeof updated === 'undefined') {
                throw new Error('Return value missing from updater');
            }

            this.data = updated;
        }

        showData(formatter?: (value: any) => any | Promise<any>) {
            let value = this.data;
            if (formatter) {
                value = awaitPromiseSync(formatter(value));
            }

            console.log(value);
        }
    };
}
