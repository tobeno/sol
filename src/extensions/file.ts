import { Constructor } from '../interfaces/util';
import { File } from '../storage/file';

export function WithFile<T extends Constructor>(base: T) {
  return class Wrapped extends base {
    file(pathOrFile: string | File): File {
      const f = pathOrFile instanceof File ? pathOrFile : new File(pathOrFile);
      f.text = String(this);
      return f;
    }
  };
}
