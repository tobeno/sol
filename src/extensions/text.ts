import { Constructor } from '../interfaces/util';
import { ClassWithText } from '../interfaces/base';
import {
    lines,
    grepLines,
    rgrepLines,
    sortLines,
    rsortLines,
    filterLines,
    rfilterLines,
    replaceLines,
    mapLines
} from '../utils/text';
import { awaitPromiseSync } from '../utils/async';

export function WithText<T extends Constructor<ClassWithText>>(base: T) {
    return class Wrapped extends base {
        grepLines(search: string | RegExp) {
            return this.updateText((data) => {
                data = grepLines(data, search);

                return data;
            });
        }

        rgrepLines(search: string | RegExp) {
            return this.updateText((data) => {
                data = rgrepLines(data, search);

                return data;
            });
        }

        sortLines() {
            return this.updateText((data) => {
                data = sortLines(data);

                return data;
            });
        }

        rsortLines() {
            return this.updateText((data) => {
                data = rsortLines(data);

                return data;
            });
        }

        filterLines(cb: (line: string) => boolean) {
            return this.updateText((data) => {
                data = filterLines(data, cb);

                return data;
            });
        }

        rfilterLines(cb: (line: string) => boolean) {
            return this.updateText((data) => {
                data = rfilterLines(data, cb);

                return data;
            });
        }

        replaceLines(pattern: string | RegExp, replacer: any) {
            return this.updateText((data) => {
                data = replaceLines(data, pattern, replacer);

                return data;
            });
        }

        mapLines(cb: (line: string) => any) {
            return mapLines(this.text, cb);
        }

        updateLines(updater: (value: string) => string | Promise<string>) {
            const updated = awaitPromiseSync(Promise.all(lines(this.text).map(updater)));
            if (typeof updated === 'undefined') {
                throw new Error('Return value missing from updater');
            }

            this.text = updated.join('\n') + '\n';

            return this.text;
        }

        updateText(updater: (value: string) => string | Promise<string>) {
            const updated = awaitPromiseSync(updater(this.text));
            if (typeof updated === 'undefined') {
                throw new Error('Return value missing from updater');
            }

            this.text = updated;

            return this.text;
        }

        showText(formatter?: (text: string) => any | Promise<any>) {
            let text = this.text;
            if (formatter) {
                text = awaitPromiseSync(formatter(text));
            }

            console.log(text);
        }
    };
}
