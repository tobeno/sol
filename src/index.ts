export { web } from './web';
export * as os from './os/fn';
export * as data from './data/fn';
export * as utils from './utils/fn';
export * as shell from './shell/fn';
export * as storage from './storage/fn';
