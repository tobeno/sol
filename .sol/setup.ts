// --- SETUP ---
/* eslint-disable */
import { globals } from '/Users/tobi/Projects/sol/src/globals';
const {
  web,
  csv,
  html,
  json,
  xml,
  ast,
  yaml,
  sol,
  astTypes,
  jsonata,
  cheerio,
  cwd,
  shell,
  utils,
  log,
  vscode,
  play,
  replay,
  unwatchPlay,
  clipboard,
  file,
  dir,
  files,
  dirs,
  grep,
  glob,
  fetch,
  test,
} = globals;
// @ts-ignore
const used = [
  web,
  csv,
  html,
  json,
  xml,
  ast,
  yaml,
  sol,
  astTypes,
  jsonata,
  cheerio,
  cwd,
  shell,
  utils,
  log,
  vscode,
  play,
  replay,
  unwatchPlay,
  clipboard,
  file,
  dir,
  files,
  dirs,
  grep,
  glob,
  fetch,
  test,
];
// @ts-ignore
const unwatch = () => {};
const shared = global as any;
let result: any = undefined;
// --- END SETUP ---
