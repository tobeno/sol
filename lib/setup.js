const path = require('path');
const fs = require('fs');

const rootPath = path.resolve(__dirname, '..');

const { register } = require('@swc-node/register/register');

const tsconfig = JSON.parse(
  fs
    .readFileSync(path.join(rootPath, 'tsconfig.json'), 'utf8')
    .replace(/(\s\/\*[^/]+?\*\/|\n *\/\/.+)/g, ''),
);

register(tsconfig.compilerOptions);

require('../src/setup');
