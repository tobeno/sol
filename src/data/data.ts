import { Expression } from 'jsonata';
import * as jsonata from 'jsonata';
import {
  wrapString,
  dataToJson,
  dataToYaml,
  dataToCsv,
  wrapObject,
} from './mapper';
import { Wrapper } from './wrapper';
import { Text } from './text';
import { DataSource } from './data-source';
import { inspect } from 'util';
import { DataTransformation } from './data-transformation';
import { clipboard } from '../os/clipboard';
import { AnyItemType, AnyKeyType, AnyPartial } from '../interfaces/util';
import { edit } from '../integrations/editor';
import { File } from '../storage/file';
import { save, saveAs } from '../storage/save';
import { camelcaseObject, snakecaseObject } from '../utils/object';

/**
 * Generic wrapper for runtime objects
 */
export class Data<
  ValueType = any,
  ItemType = AnyItemType<ValueType>,
  KeyType = AnyKeyType<ValueType>
> extends Wrapper<ValueType> {
  constructor(
    value: ValueType,
    public source: DataSource | null = null,
    public sourceTransformation: DataTransformation | null = null,
  ) {
    super(value);
  }

  get rootSource(): DataSource | null {
    if (!this.source) {
      return null;
    }

    return this.source.rootSource || this.source;
  }

  get text(): Text<ValueType> {
    let value = this.value as any;
    if (!(value instanceof Text) && typeof value !== 'string') {
      value = this.json;
    }

    return wrapString(value, null, this);
  }

  get json(): Text<ValueType> {
    return dataToJson<ValueType>(this);
  }

  get yaml(): Text<ValueType> {
    return dataToYaml<ValueType>(this);
  }

  get csv(): Text<ValueType> {
    return dataToCsv<ValueType>(this);
  }

  get clone(): Data<ValueType> {
    return wrapObject<ValueType>(JSON.parse(JSON.stringify(this.value)), this);
  }

  get flattened(): Data<
    ValueType extends Array<any> ? FlatArray<ValueType, 1> : ValueType
  > {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.flat() as any, this) as any;
    }

    return this as any;
  }

  get unique(): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return wrapObject([...new Set(this.value)] as any, this) as any;
    }

    const knownValues: any[] = [];

    return this.filter((value: any) => {
      const result = !knownValues.includes(value);
      knownValues.push(value);
      return result;
    });
  }

  get sorted(): Data<ValueType extends Array<any> ? ValueType : any> {
    return this.sort();
  }

  get filtered(): Data<AnyPartial<ValueType>> {
    return this.filter((value: any) => !!value);
  }

  get camelcased(): Data<any> {
    return wrapObject(camelcaseObject(this.value), this) as any;
  }

  get snakecased(): Data<any> {
    return wrapObject(snakecaseObject(this.value), this) as any;
  }

  get keys(): KeyType[] {
    if (Array.isArray(this.value)) {
      return (this.value as any).map((_: any, index: any) => index) as any;
    }

    return Object.keys(this.value) as any;
  }

  get values(): ItemType[] {
    if (Array.isArray(this.value)) {
      return this.value as any;
    }

    return Object.values(this.value);
  }

  get entries(): [KeyType, ItemType][] {
    if (Array.isArray(this.value)) {
      return this.value.entries() as any;
    }

    return Object.entries(this.value) as any;
  }

  withSource(source: DataSource | null): Data<ValueType> {
    this.source = source;

    return this as any;
  }

  sort(
    compareFn?: (a: ItemType, b: ItemType) => number,
  ): Data<ValueType extends Array<any> ? ValueType : any> {
    if (Array.isArray(this.value)) {
      return wrapObject([...this.value].sort(compareFn), this) as any;
    }

    const entries = this.entries;
    entries.sort(([aIndex, a], [bIndex, b]) => {
      if (compareFn) {
        return compareFn(a, b);
      }

      if (a < b) {
        return -1;
      }

      if (a > b) {
        return 1;
      }

      return 0;
    });

    return wrapObject(Object.fromEntries(entries), this) as any;
  }

  filter(
    cb: (value: ItemType, index: KeyType) => boolean,
  ): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.filter(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        newValue[index] = value;
      }
    });

    return wrapObject(newValue, this) as any;
  }

  find(cb: (value: ItemType, index: KeyType) => boolean): Data<ItemType> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.find(cb as any) as any, this) as any;
    }

    let result: any = null;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = value;

        break;
      }
    }

    return result ? (wrapObject(result, this) as any) : null;
  }

  findIndex(cb: (value: ItemType, index: KeyType) => boolean): Data<KeyType> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.findIndex(cb as any) as any, this) as any;
    }

    let result: any = null;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = index;

        break;
      }
    }

    return result ? (wrapObject(result, this) as any) : null;
  }

  some(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.some(cb as any);
    }

    let result = false;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = true;

        break;
      }
    }

    return result;
  }

  every(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.every(cb as any);
    }

    let result = true;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (!cb(value, index)) {
        result = false;

        break;
      }
    }

    return result;
  }

  map<MappedItemType = any>(
    cb: (value: ItemType, index: KeyType) => MappedItemType,
  ): Data<
    ValueType extends Array<any>
      ? MappedItemType[]
      : ValueType extends Record<string, any>
      ? { [key: string]: MappedItemType }
      : any
  > {
    if (Array.isArray(this.value)) {
      return wrapObject((this.value as any).map(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      newValue[index] = cb(value, index);
    });

    return wrapObject(newValue, this) as any;
  }

  join(separator: string): Text {
    if (Array.isArray(this.value)) {
      return wrapString(this.value.join(separator), null, this);
    }

    return wrapString(Object.values(this.value).join(separator), null, this);
  }

  forEach(cb: (value: ItemType, index: KeyType) => any) {
    if (Array.isArray(this.value)) {
      return this.value.forEach(cb as any);
    }

    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      cb(value, index);
    });
  }

  transform<TransformedValueType = any>(
    exp: string | Expression,
  ): Data<TransformedValueType> {
    if (typeof exp === 'string') {
      exp = jsonata(exp);
    }

    return wrapObject(exp.evaluate(this.value), this) as any;
  }

  copy() {
    clipboard.text = String(this);
  }

  edit(): File {
    return edit(this);
  }

  print() {
    console.log(String(this));
  }

  saveAs(path: string): File {
    return saveAs(this, path);
  }

  save(): File {
    return save(this);
  }

  [inspect.custom]() {
    return `Data ${inspect(this.value)}`;
  }
}
