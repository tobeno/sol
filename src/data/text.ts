import {
  csvToData,
  yamlToData,
  jsonToData,
  wrapObject,
  wrapString,
} from './transformer';
import {
  lines,
  grepLines,
  rgrepLines,
  sortLines,
  rsortLines,
  filterLines,
  rfilterLines,
  replaceLines,
  mapLines,
  extractText,
} from '../utils/text';
import { clipboard } from '../os/clipboard';
import { edit } from '../integrations/editor';
import { DataFormat } from './data-format';
import { DataSource } from './data-source';
import { DataTransformation } from './data-transformation';
import { File } from '../storage/file';
import { inspect } from 'util';
import { Data } from './data';
import { save, saveAs } from '../storage/save';

/**
 * Wrapper for strings
 */
export class Text<ContentType = any> extends String {
  constructor(
    value: string | String,
    public format: string | null = null,
    public source: DataSource | null = null,
    public sourceTransformation: DataTransformation | null = null,
  ) {
    super(value.toString());
  }

  get value(): string {
    return this.toString();
  }

  get rootSource(): DataSource | null {
    if (!this.source) {
      return null;
    }

    return this.source.rootSource || this.source;
  }

  get ext(): string {
    return DataFormat.toExt(this.format);
  }

  get text(): this {
    return this;
  }

  get lines(): Data<string[]> {
    return wrapObject(lines(this.toString()), this);
  }

  get json(): Data<ContentType> {
    return jsonToData(this);
  }

  get yaml(): Data<ContentType> {
    return yamlToData(this);
  }

  get csv(): Data<ContentType> {
    return csvToData(this);
  }

  withFormat(format: string | null): this {
    this.format = format;

    return this;
  }

  withSource(source: DataSource | null): this {
    this.source = source;

    return this;
  }

  grepLines(search: string | RegExp): Text<ContentType> {
    return wrapString(grepLines(String(this), search), this.format, this);
  }

  rgrepLines(search: string | RegExp): Text<ContentType> {
    return wrapString(rgrepLines(String(this), search), this.format, this);
  }

  sortLines(): Text<ContentType> {
    return wrapString(sortLines(String(this)), this.format, this);
  }

  rsortLines(): Text<ContentType> {
    return wrapString(rsortLines(String(this)), this.format, this);
  }

  filterLines(cb: (line: string) => boolean): Text<ContentType> {
    return wrapString(filterLines(String(this), cb), this.format, this);
  }

  rfilterLines(cb: (line: string) => boolean): Text<ContentType> {
    return wrapString(rfilterLines(String(this), cb), this.format, this);
  }

  replaceLines(pattern: string | RegExp, replacer: any): Text<ContentType> {
    return wrapString(
      replaceLines(String(this), pattern, replacer),
      this.format,
      this,
    );
  }

  mapLines(cb: (line: string) => any): Text<ContentType> {
    return wrapString(mapLines(String(this), cb), this.format, this);
  }

  extract(pattern: string | RegExp): Data<Text[]> {
    return wrapObject(
      extractText(String(this), pattern).map((s) => wrapString(s)),
      this,
    );
  }

  print(): void {
    console.log(this.toString());
  }

  edit(): File {
    return edit(this);
  }

  eval(): any {
    return eval(this.toString());
  }

  copy(): void {
    clipboard.text = this;
  }

  saveAs(path: string): File {
    return saveAs(this, path);
  }

  save(): File {
    return save(this);
  }

  [inspect.custom]() {
    return this.toString();
  }
}
