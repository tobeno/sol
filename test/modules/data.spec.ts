import '../../src/setup';
import tmp from 'tmp';

import { File } from '../../src/modules/storage/file';

import { readFileSync } from 'fs';
import { Data } from '../../src/modules/data/data';
import { Text } from '@sol/modules/data/text';

interface ProductVariant {
  id: string;
  size?: string;
  quantity: number;
}

interface Product {
  id: string;
  name: string;
  variants: ProductVariant[];
}

interface ProductsFile {
  products: Product[];
}

describe('data module', () => {
  describe('transformation', () => {
    describe('file', () => {
      it('should transform a JSON file', async () => {
        const data = File.create<ProductsFile>(
          `${__dirname}/../assets/products.json`,
        ).json;
        const obj = data.value;

        expect(obj).toHaveProperty('products');

        (obj as any).variants = Array.from(
          Data.create(obj.products).extract('variants').value,
        );

        const tmpFile = tmp.fileSync();
        data.json.saveAs(tmpFile.name);
        expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
        tmpFile.removeCallback();

        const tmpFile2 = tmp.fileSync();
        data.extract('variants').csv.saveAs(tmpFile2.name);
        expect(readFileSync(tmpFile2.name, 'utf-8')).toMatchSnapshot();
        tmpFile2.removeCallback();
      });

      it('should transform a YAML file', async () => {
        const data = File.create<ProductsFile>(
          `${__dirname}/../assets/products.yaml`,
        ).yaml;
        const obj = data.value;

        expect(obj).toHaveProperty('products');

        const tmpFile = tmp.fileSync();
        data.yaml.saveAs(tmpFile.name);
        expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
        tmpFile.removeCallback();
      });

      it('should transform a CSV file', async () => {
        let data = File.create<ProductVariant[]>(
          `${__dirname}/../assets/variants.csv`,
        ).csv;

        data = data.sort((a, b) => a.quantity - b.quantity);

        const tmpFile = tmp.fileSync();
        data.csv.saveAs(tmpFile.name);
        expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
        tmpFile.removeCallback();
      });
    });

    describe('object', () => {
      it('should transform an array', async () => {
        let jsonData = Data.create([{ id: '1' }, { id: '2' }] as {
          id: string;
        }[]);

        expect(jsonData.json.value).toMatchSnapshot();

        const filteredData = jsonData.filter((item) => item.id === '1');

        const mappedData = filteredData.map((item) => item.id);

        expect(mappedData.value).toEqual(['1']);
      });

      it('should transform an object', async () => {
        let data = Data.create({
          'product-1': { name: 'Shoe' },
          'product-2': { name: 'Shirt' },
        } as Record<string, { name: string }>);

        expect(data.json.value).toMatchSnapshot();

        const filteredData = data.filter((item) => item.name === 'Shoe');

        const mappedData = filteredData.map((item) => item.name);

        expect(mappedData.value).toEqual({ 'product-1': 'Shoe' });
      });

      it('should transform a string', async () => {
        let text = Text.create(
          'Product 1: Shoe\nProduct 2: Shirt\nProduct 3: Jeans',
        );

        let data = text.lines
          .map((line) => {
            const match = line.match(/Product (\d+): (.+)/);
            if (!match) {
              return null;
            }

            return { id: match[0], name: match[1] };
          })
          .filter((item) => !!item)
          .map((item) => JSON.stringify(item));

        text = data.join('.');

        expect(text.value).toMatchSnapshot();
      });
    });
  });
});
