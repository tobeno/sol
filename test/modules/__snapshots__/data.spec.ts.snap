// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data module transformation file should transform a CSV file 1`] = `
"id,size,quantity
1-3,44,1
2-3,38,1
2-1,36,2
2-2,37,2
1-1,42,3
1-2,43,20"
`;

exports[`data module transformation file should transform a JSON file 1`] = `
"{
  \\"products\\": [
    {
      \\"id\\": \\"1\\",
      \\"name\\": \\"Leather shoe\\",
      \\"variants\\": [
        {
          \\"id\\": \\"1-1\\",
          \\"size\\": \\"42\\",
          \\"quantity\\": 3
        },
        {
          \\"id\\": \\"1-2\\",
          \\"size\\": \\"43\\",
          \\"quantity\\": 20
        },
        {
          \\"id\\": \\"1-3\\",
          \\"size\\": \\"44\\",
          \\"quantity\\": 1
        }
      ]
    },
    {
      \\"id\\": \\"2\\",
      \\"name\\": \\"Stiletto shoe\\",
      \\"variants\\": [
        {
          \\"id\\": \\"2-1\\",
          \\"size\\": \\"36\\",
          \\"quantity\\": 2
        },
        {
          \\"id\\": \\"2-2\\",
          \\"size\\": \\"37\\",
          \\"quantity\\": 2
        },
        {
          \\"id\\": \\"2-3\\",
          \\"size\\": \\"38\\",
          \\"quantity\\": 1
        }
      ]
    }
  ],
  \\"variants\\": [
    {
      \\"id\\": \\"1-1\\",
      \\"size\\": \\"42\\",
      \\"quantity\\": 3
    },
    {
      \\"id\\": \\"1-2\\",
      \\"size\\": \\"43\\",
      \\"quantity\\": 20
    },
    {
      \\"id\\": \\"1-3\\",
      \\"size\\": \\"44\\",
      \\"quantity\\": 1
    },
    {
      \\"id\\": \\"2-1\\",
      \\"size\\": \\"36\\",
      \\"quantity\\": 2
    },
    {
      \\"id\\": \\"2-2\\",
      \\"size\\": \\"37\\",
      \\"quantity\\": 2
    },
    {
      \\"id\\": \\"2-3\\",
      \\"size\\": \\"38\\",
      \\"quantity\\": 1
    }
  ]
}"
`;

exports[`data module transformation file should transform a JSON file 2`] = `
"id,size,quantity
1-1,42,3
1-2,43,20
1-3,44,1
2-1,36,2
2-2,37,2
2-3,38,1"
`;

exports[`data module transformation file should transform a YAML file 1`] = `
"products:
  - id: \\"1\\"
    name: Leather shoe
    variants:
      - id: 1-1
        size: \\"42\\"
        quantity: 3
      - id: 1-2
        size: \\"43\\"
        quantity: 20
      - id: 1-3
        size: \\"44\\"
        quantity: 1
  - id: \\"2\\"
    name: Stiletto shoe
    variants:
      - id: 2-1
        size: \\"36\\"
        quantity: 2
      - id: 2-2
        size: \\"37\\"
        quantity: 2
      - id: 2-3
        size: \\"38\\"
        quantity: 1
"
`;

exports[`data module transformation object should transform a string 1`] = `"{\\"id\\":\\"Product 1: Shoe\\",\\"name\\":\\"1\\"}.{\\"id\\":\\"Product 2: Shirt\\",\\"name\\":\\"2\\"}.{\\"id\\":\\"Product 3: Jeans\\",\\"name\\":\\"3\\"}"`;

exports[`data module transformation object should transform an array 1`] = `
"[
  {
    \\"id\\": \\"1\\"
  },
  {
    \\"id\\": \\"2\\"
  }
]"
`;

exports[`data module transformation object should transform an object 1`] = `
"{
  \\"product-1\\": {
    \\"name\\": \\"Shoe\\"
  },
  \\"product-2\\": {
    \\"name\\": \\"Shirt\\"
  }
}"
`;
